receiver/dotnetdiagnosticsreceiver/dotnet/request.go

```diff
@@ -0,0 +1,240 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+// RequestWriter submits the initial request to the dotnet diagnostics backend,
+// after which it starts sending us metrics at the specified interval.
+// ByteBuffer is swappable for testing.
+type RequestWriter struct {
+	buf         network.ByteBuffer
+	w           io.Writer
+	intervalSec int
+	names       []string
+}
+
+func NewRequestWriter(buf network.ByteBuffer, w io.Writer, intervalSec int, names ...string) *RequestWriter {
+	return &RequestWriter{buf: buf, w: w, intervalSec: intervalSec, names: names}
+}
+
+func (w *RequestWriter) SendRequest() error {
+	req, err := w.createRequest()
+	if err != nil {
+		return err
+	}
+	_, err = w.w.Write(req)
+	return err
+}
+
+const eventPipeCommand = 2
+const collectTracing2CommandID = 3
+
+func (w *RequestWriter) createRequest() ([]byte, error) {
+	cfgReq := newConfigRequest(w.intervalSec, w.names...)
+	payload, err := cfgReq.serialize(w.buf)
+	if err != nil {
+		return nil, err
+	}
+
+	hdr := &requestHeader{
+		commandSet: eventPipeCommand,
+		commandID:  collectTracing2CommandID,
+	}
+	hdrBytes, err := hdr.serialize(w.buf, len(payload))
+	if err != nil {
+		return nil, err
+	}
+
+	return append(hdrBytes, payload...), nil
+}
+
+// requestHeader
+type requestHeader struct {
+	commandSet byte
+	commandID  byte
+}
+
+const magic = "DOTNET_IPC_V1"
+const requestHeaderSize = 20
+
+func (h requestHeader) serialize(buf network.ByteBuffer, payloadSize int) ([]byte, error) {
+	buf.Reset()
+
+	_, err := buf.Write([]byte(magic + "\000"))
+	if err != nil {
+		return nil, err
+	}
+
+	totSize := uint16(requestHeaderSize + payloadSize)
+	err = binary.Write(buf, network.ByteOrder, totSize)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = buf.Write([]byte{h.commandSet, h.commandID, 0, 0})
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]byte, buf.Len())
+	copy(out, buf.Bytes())
```

Comment:
Do we need to create a copy here or can use Bytes() directly? Is the `buf` going to be used for some other purpose after this? Question: why don't we create a new local `buf` in this func every time? It would be same/similar number of allocations but would not require the copy anymore, I think.

Reply:
Originally, I just created a new buf every time I needed one, like you suggest, but I realized that I would have to mock failures to get code coverage up to an acceptable level -- there were lots of `if err != nil` blocks not covered. So I had to decide whether to pass in a byte buffer creator that could be swapped for testing, or pass in a single byte buffer instance that could be swapped for testing and `Reset()` it plus copy its results (necessary because `buf.Bytes()` returns a view of the backing array, which is overwritten later). I chose the latter because it seemed to make the code easier to read. I don't have a strong opinion on this matter though and am open to suggestions.

Reply:
Update: It turns out that bytes.Buffer always returns a nil error, so all of the error handling and complexity around testing the error blocks was unnecessary. I've removed it and reverted to just creating a new bytes.Buffer when necessary.